src/Components/Components/src/Routing/SupplyParameterFromQueryValueProvider.cs

```diff
@@ -0,0 +1,134 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using Microsoft.AspNetCore.Components.Rendering;
+
+namespace Microsoft.AspNetCore.Components.Routing;
+
+internal sealed class SupplyParameterFromQueryValueProvider(NavigationManager navigationManager) : ICascadingValueSupplier, IDisposable
+{
+    private QueryParameterValueSupplier? _queryParameterValueSupplier;
+    private HashSet<ComponentState>? _subscribers;
+    private HashSet<ComponentState>? _pendingSubscribers;
+    private bool _isSubscribedToLocationChanges;
+    private string? _lastUri;
+
+    public bool IsFixed => false;
+
+    public bool CanSupplyValue(in CascadingParameterInfo parameterInfo)
+        => parameterInfo.Attribute is SupplyParameterFromQueryAttribute;
+
+    public object? GetCurrentValue(in CascadingParameterInfo parameterInfo)
+    {
+        TryUpdateQueryParameters();
+
+        var attribute = (SupplyParameterFromQueryAttribute)parameterInfo.Attribute; // Must be a valid cast because we check in CanSupplyValue
+        var queryParameterName = attribute.Name ?? parameterInfo.PropertyName;
+        return _queryParameterValueSupplier.GetQueryParameterValue(parameterInfo.PropertyType, queryParameterName);
+    }
+
+    public void Subscribe(ComponentState subscriber, in CascadingParameterInfo parameterInfo)
+    {
+        if (_pendingSubscribers?.Count > 0 || (TryUpdateQueryParameters() && _isSubscribedToLocationChanges))
+        {
+            // This branch is only taken if there's a pending OnLocationChanged event for the current Uri that we're already subscribed to.
+            // We'll add the _pendingSubscribers to _subscribers and clear _pendingSubscribers during the upcoming call to OnLocationChanged.
+            _pendingSubscribers ??= new();
+            _pendingSubscribers.Add(subscriber);
+            return;
+        }
+
+        _subscribers ??= new();
+        _subscribers.Add(subscriber);
+        SubscribeToLocationChanges();
+    }
+
+    public void Unsubscribe(ComponentState subscriber, in CascadingParameterInfo parameterInfo)
```

Comment:
One thing I do not like about how this works is that the same `ComponentState` can subscribe multiple times in a row for multiple different parameters, but as soon as it unsubscribes one parameter, it gets completely unsubscribed.

However, I don't know how a component would unsubscribe from a subset of parameters in practice, and this was a preexisting problem with the class. Do we care about fixing this?

Reply:
We could just change the sets to be `HashSet<(ComponentState Subscriber, string QueryName)>` rather than use a dictionary if we think this is important. I just wanted to make sure I wasn't fixing something that wasn't really broken. While the current behavior is certainly very weird which is why I pointed it out, I was worried "fixing" it in this way could lead to excessive parameter change notifications and renders when a component has multiple query string parameters.

> Or perhaps even when navigating away from any page that subscribed to a query parameter, possibly causing extra rendering work.

I was worried about this too, but this fortunately does not induce extra rendering work because `Router.OnLocationChanged` gets called first and unsubscribes clearing the `_subcribers` before `SupplyParameterFromQueryValueProvider.OnLocationChanged` has a chance to notify them.

Reply:
> We could just change the sets to be HashSet<(ComponentState Subscriber, string QueryName)> rather than use a dictionary if we think this is important.

I thought we needed to be able to get the set of distinct subscribers, since at notification time, we only want to issue one notification per subscriber even if multiple query params have changed (a common case). The tuple would force us to compute the set of distinct subscribers on each notification.

> I was worried about this too, but this fortunately does not induce extra rendering work 

In that case it seems fine already :)